### PR TITLE
[RLlib] New ConnectorV2 API #06: Changes in SingleAgentEpisode & SingleAgentEnvRunner.

### PR DESCRIPTION
**Pull request by sven1977**
This PR adds some changes to SingleAgentEpisode & SingleAgentEnvRunner:

* SingleAgentEnvRunner now utilizes the user-configured EnvToModule and ModuleToEnv connector pipelines.
* Hence, `SingleAgentEnvRunner` does NOT anymore:
* * Have to deal with internal (RNN) states.
* * Action sampling, logp, probs computations as these are all done automatically in the default ModuleToEnv pipeline.

* Add `set` APIs to SingleAgentEpisode, such that custom connectors are able to manipulate an episode's data, e.g. for observation framestacking, reward clipping, etc..
* New `set` API had to also be supported then by `InfiniteLookbackBuffer`, which sits at the core of all episode classes.

* Updated test cases and added new ones for `set` APIs.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/single_agent_env_runner.py [hunk 1]
```diff
@@ -623,40 +559,6 @@ def stop(self):
         # Close our env object via gymnasium's API.
         self.env.close()
 
-    # TODO (sven): Replace by default "to-env" connector.
```
Comment (sven1977) [inline rllib/env/single_agent_env_runner.py:626]: Not necessary anymore here in EnvRunner.

This is default ModuleToEnv connector behavior now.

File: rllib/env/single_agent_env_runner.py [hunk 2]
```diff
@@ -221,111 +217,72 @@ def _sample_timesteps(
...
-                }
-                from ray.rllib.utils.nested_dict import NestedDict
-
-                batch = NestedDict(batch)
-
+                to_module = self._env_to_module(
+                    rl_module=self.module,
+                    episodes=self._episodes,
+                    explore=explore,
+                    # persistent_data=None, #TODO
```
Comment (kouroshHakha) [inline rllib/env/single_agent_env_runner.py:236]: Is there a TODO here? What is the todo exactly?

File: rllib/env/single_agent_env_runner.py [hunk 3]
```diff
@@ -221,111 +217,72 @@ def _sample_timesteps(
...
-                # TODO (sven): Will be completely replaced by connector logic in
-                #  upcoming PR.
-                actions, action_logp = self._sample_actions_if_necessary(
-                    fwd_out, explore
+                to_env = self._module_to_env(
+                    rl_module=self.module,
+                    data=to_env,
+                    episodes=self._episodes,
+                    explore=explore,
+                    # persistent_data=None, #TODO
```
Comment (kouroshHakha) [inline rllib/env/single_agent_env_runner.py:249]: same thing

**Review by kouroshHakha — APPROVED** [review]

File: rllib/connectors/learner/default_learner_connector.py [hunk 4]
```diff
@@ -161,6 +161,15 @@ def __call__(
...
 
+        # TODO (sven): Convert data to proper tensor formats, depending on framework
+        #  used by the RLModule. We cannot do this right now as the RLModule does NOT
+        #  know its own device. Only the Learner knows the device. Also, on the
+        #  EnvRunner side, we assume that it's always the CPU (even though one could
+        #  imagine a GPU-based EnvRunner + RLModule for sampling).
+        # if rl_module.framework == "torch":
+        #    data = convert_to_torch_tensor(data, device=??)
+        # elif rl_module.framework == "tf2":
+        #    data =
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/connectors/learner/default_learner_connector.py:172]: It's a TODO on an open question with the possible code-solution commented out. I'll leave this in. We need to unify this behavior (numpy to tensor) for all connector types in the near future to not cause user confusion.
The blocker right now is the fact that an RLModule does not know its own device today (only Learners do (GPU or CPU) and EnvRunners assume they are always on the CPU). Thus, connectors have to means to perform this conversion step properly.

**Pull request merged**